src/fenicsxconcrete/bcs.py

```diff
@@ -0,0 +1,36 @@
+
+from pydantic import ConfigDict, RootModel
+from pydantic.dataclasses import dataclass
+from pydantic.types import conlist
+from typing import Callable, NewType
+import numpy as np
+
+Marker = NewType('Marker', int)
+
+@dataclass(config=dict(arbitrary_types_allowed=True))
+class DirichletBCDefinition:
+    """
+    Definition of a time- and position-dependent Dirichlet boundary condition.
+    Note that functions cannot be serialized to JSON!
+    """
+    marker: Marker | list[float] | Callable[[np.ndarray], np.ndarray]
+    value: conlist(float, min_length=1, max_length=3) | Callable[[np.ndarray, float], np.ndarray]
+    subspace: int | None
+    variable: str
+
+@dataclass(config=dict(arbitrary_types_allowed=True))
+class NeumannBCDefinition:
+    """
+    Definition of a time- and position-dependent Neumann boundary condition.
+    """
+    marker: Marker
+    value: conlist(float, min_length=1, max_length=3) | Callable[[np.ndarray, float], np.ndarray] 
+
+@dataclass(config=dict(arbitrary_types_allowed=True))
+class BodyForceDefinition:
```

Comment:
Yeah, I wasn't sure what to do here. What would be the general term for _body force_? So any integral

$$
\int_\Omega \mathbf b \rho \mathrm d x
$$

Reply:
source term?

Reply:
I am not sure how we should generalize this. As far as I see it, the body-force definition is constant because gravity is constant. But how is it with heat flux? From this doc https://www.simscale.com/docs/simulation-setup/boundary-conditions/volume-heat-flux/ I get that it could vary with time and position

Reply:
wouldnt't that be the same as the functional representation of boundary conditions being a function of space and time? So there could certainly be constant or linear functions as standard (both relevant for body forces or linearly increasing boundary conditions) and then have the functional representation as additional option.

Reply:
True, but my question is more related to what should be allowed. In an ideal world, the data structure would ensure that only valid values (like constant gravity) are allowed, but I guess that should then be the responsibility of the user.

Reply:
Well, I guess it will be very hard to be able to define problems and test the validity beforehand (just from the syntax). And for Dirichlet BC I see now way around that, otherwise we are very limited, so then it makes no difference it that is added here again. But we could define two functions, one for BodyForce = constant and SpaceTimeDependentBodyForce(x,t).